opentimelineio_contrib/adapters/xges.py

```diff
@@ -0,0 +1,774 @@
+                META_NAMESPACE: {
+                    "metadatas": timeline.attrib.get("metadatas", "metadatas"),
+                    "properties": timeline.attrib.get("properties", "properties")
+                }
+            }
+        )
+
+        all_names = set()
+        self._add_layers(timeline, otio_timeline, all_names)
+        otio_project.append(otio_timeline)
+
+        return otio_project
+
+    def _add_layers(self, timeline, otio_timeline, all_names):
+        for layer in timeline.findall("./layer"):
+            tracks = self._build_tracks_from_layer_clips(layer, all_names)
+            otio_timeline.tracks.extend(tracks)
+
+    def _get_clips_for_type(self, clips, track_type):
+        if not clips:
+            return False
+
+        clips_for_type = []
+        for clip in clips:
+            if int(clip.attrib['track-types']) & track_type:
+                clips_for_type.append(clip)
+
+        return clips_for_type
+
+    def _build_tracks_from_layer_clips(self, layer, all_names):
+        all_clips = layer.findall('./clip')
+
+        tracks = []
+        for track_type in [GESTrackType.VIDEO, GESTrackType.AUDIO]:
+            clips = self._get_clips_for_type(all_clips, track_type)
+            if not clips:
+                continue
+
+            track = otio.schema.Track()
+            track.kind = GESTrackType.to_otio_type(track_type)
+            self._add_clips_in_track(clips, track, all_names)
+
+            tracks.append(track)
+
+        return tracks
+
+    def _add_clips_in_track(self, clips, track, all_names):
+        for clip in clips:
+            otio_clip = self._create_otio_clip(clip, all_names)
+            if otio_clip is None:
+                continue
+
+            clip_offset = self.to_rationnal_time(int(clip.attrib['start']))
+            if clip_offset > track.duration():
+                track.append(
+                    self._create_otio_gap(
+                        0,
+                        (clip_offset - track.duration())
+                    )
+                )
+
+            track.append(otio_clip)
+
+        return track
+
+    def _get_clip_name(self, clip, all_names):
+        i = 0
+        tmpname = name = clip.get("name", GstStructure(clip.get("properties", "properties;")).get("name"))
+        while True:
+            if tmpname not in all_names:
+                all_names.add(tmpname)
+                return tmpname
+
+            i += 1
+            tmpname = name + '_%d' % i
+
+    def _create_otio_transition(self, clip, all_names):
+        start = self.to_rationnal_time(clip.attrib["start"])
+        end = start + self.to_rationnal_time(clip.attrib["duration"])
+        cut_point = otio.opentime.RationalTime((end.value - start.value) / 2, start.rate)
+
+        return otio.schema.Transition(
+            name=self._get_clip_name(clip, all_names),
+            transition_type=TRANSITION_MAP.get(
+                clip.attrib["asset-id"], otio.schema.TransitionTypes.Custom
```

Comment:
It looks like GStreamer has a nice set of transition types. I like how you've made a map here so this can be easily extended once OTIO supports more types. Does GStreamer support anything beyond the SMPTE transition types?

Reply:
Currently the GES.VideoTransition only supports SMPTE but we could use effects (and the huge database of effects elements from GStreamer) to basically create other types.